pil-analyzer/src/pil_analyzer.rs

```diff
@@ -257,6 +258,64 @@ impl PILAnalyzer {
         }
     }
 
+    /// Verifies that all struct instantiations match their corresponding declarations (existence of field names, completeness)
+    fn struct_fields_check(&self) -> Result<(), Vec<Error>> {
+        let mut errors = vec![];
+
+        self.all_children().for_each(|expr| {
+            if let Expression::StructExpression(sr, struct_expr) = expr {
+                let Some((
+                    _,
+                    Some(FunctionValueDefinition::TypeDeclaration(TypeDeclaration::Struct(
+                        struct_decl,
+                    ))),
+                )) = self.definitions.get(&struct_expr.name.to_string())
+                else {
+                    panic!("Struct '{}' has not been declared", struct_expr.name);
+                };
+
+                let mut declared_fields = HashSet::new();
+                for field in &struct_decl.fields {
+                    if !declared_fields.insert(&field.name) {
+                        errors.push(
+                            sr.with_error(format!("Field '{}' is already declared", field.name)),
```

Comment:
Oh I see where this error message comes from. If you do it this way, then 1) we won't get an error for a struct declaration that is never used and 2) we will get the error at the location of the use and not the declaration and 3) we will get multiple errors for each use.